src/core/qgscoordinatetransform.cpp

```diff
@@ -472,6 +474,14 @@ void QgsCoordinateTransform::transformCoords( int numPoints, double *x, double *
 
   // use proj4 to do the transform
 
+#ifdef QGISDEBUG
+  // thread safety check - if this fails, you haven't called detachForThread()!
+  if ( !d->mOwnerThread )
+    d->mOwnerThread = QThread::currentThread();
+  else
+    Q_ASSERT( QThread::currentThread() == d->mOwnerThread );
```

Comment:
To be really nice this could be replaced with

    Q_ASSERT_X( QThread::currentThread() == d->mOwnerThread, "transformCoords", "Using `QgsCoordinateTransform` in background thread without calling `detachForThread()`");